website/blog/2023-06-01-ariadne-codegen-0-7.md

```diff
@@ -0,0 +1,307 @@
+
+For example, given following operation:
+
+```gql
+subscription GetUsersCounter {
+    usersCounter
+}
+```
+
+Generated client will have following method:
+
+```py
+    async def get_users_counter(self) -> AsyncIterator[GetUsersCounter]:
+        query = gql(
+            """
+            subscription GetUsersCounter {
+              usersCounter
+            }
+            """
+        )
+        variables: dict[str, object] = {}
+        async for data in self.execute_ws(query=query, variables=variables):
+            yield GetUsersCounter.parse_obj(data)
+```
+
+Our default async base client uses [websockets](https://github.com/python-websockets/websockets) package and implements [graphql-transport-ws](https://github.com/enisdenjo/graphql-ws/blob/master/PROTOCOL.md) subprotocol.
+
+Required dependencies can by installed with pip:
+
+```
+$ pip install ariadne-codegen[subscriptions]
+```
+
+
+## Fragments
+
+In previous versions of Codegen fragments were "unpacked" in queries. For example, given following operations:
+
+```gql
+query GetA {
+    GetTypeA {
+        ...FragmentA
+    }
+}
+
+query ListA {
+    ListTypeA {
+        ...FragmentA
+    }
+}
+
+fragment FragmentA on TypeA {
+    id
+    name
+}
+```
+
+Generated `get_a.py` and `list_a.py` files had types looking like this:
+
+```py
+# get_a.py
+
+class GetA(BaseModel):
+    get_type_a: "GetAGetTypeA" = Field(alias="GetTypeA")
+
+
+class GetAGetTypeA(BaseModel):
+    id: str
+    name: str
+```
+
+```py
+# list_a.py
+
+class ListA(BaseModel):
+    list_type_a: List["ListAListTypeA"] = Field(alias="ListTypeA")
+
+
+class ListAListTypeA(BaseModel):
+    id: str
+    name: str
+```
+
+Both of these operations use the same `FragmentA` to represnt `TypeA`, but our models didn't show that.
+
+To make working with fragment easier, in Ariadne Codegen 0.7 we are changing this behaviour. Instead of unpacking fragments, we generate separate models for them and use them as mixins. From the same operation now we will get 3 files: `get_a.py`, `list_a.py` and `fragments.py`
+
+```py
+# get_a.py
+
+class GetA(BaseModel):
+    get_type_a: "GetAGetTypeA" = Field(alias="GetTypeA")
+
+
+class GetAGetTypeA(FragmentA):
+    pass
+```
+
+```py
+# list_a.py
+
+class ListA(BaseModel):
+    list_type_a: List["ListAListTypeA"] = Field(alias="ListTypeA")
+
+
+class ListAListTypeA(FragmentA):
+    pass
+```
+
+```py
+# fragments.py
+
+class FragmentA(BaseModel):
+    id: str
+    name: str
+```
+
+With this change you can use fragments to implement logic for multiple types, eg. `def process_a(a: FragmentA)...`. New `fragments.py` consists fragments collected from all parsed operations.
+
+### Unions and Interfaces
+
+There is exception from new fragments behaviour, if fragment represents `Union` then we unpack it as before. Example of such fragment:
+
+```gql
+fragment AnimalData on AnimalInterface {
+    name
+    ... on Dog {
+        dogField
+    }
+    ... on Cat {
+        catField
+    }
+}
+```
```

Comment:
We should add the "For above fragment this Python code will be generated:" part to show old behavior for those not familiar with it.